docs/proposals/odo-deploy.md

```diff
@@ -0,0 +1,152 @@
+
+It is not meant to replace GitOps/pipeline-based deployments for governed environments (test, stage, production). However, both `odo deploy` and `odo pipelines` must honour the build and deployment information provided by the application stack (devfile).
+
+## User Flow
+
+This command will allow a user to perform inner-loop and then test the outer-loop, so the application is truly ready for checking into git and for pipelines to take over. Here's how a typical application development flow might look like: 
+
+User flow: 
+1. `odo create <component> <mycomponent>` - This initializes odo component in the current directory.
+1. Edit the project source code to develop the application.
+1. `odo URL create` - This stores URL information (host, port etc.) for accessing the application on the cluster (if not done already).
+1. `odo push` - This runs the application source code using inner-loop instructions from devfile.
+1. Validate the running application is working as intended.
+1. Iterate over steps 2 and beyond (as needed). 
+1. `odo deploy` - This would build a new clean image and deploy it to the target cluster using outer-loop instructions from devfile and user-provided arguments.
+1. Validate the deployed application is working as intended.
+1. Iterate over steps 2 and beyond (as needed).
+1. Optionally, run `odo deploy delete` to clean up resources created with `odo deploy`.
+1. Push your code to Git - ready for sharing with your team and for CI/CD pipelines to take over.
+
+## User Stories
+
+### Initial build and deploy support to odo - https://github.com/openshift/odo/issues/3300
+
+## Design overview
+`odo deploy` could provide developers with a way to build a container image for their application and deploy it on a target Kubernetes deployment using the build/deploy guidance provided by the devfile.
+
+This deployment is equivalent to a development version of your production and will be using the namespace and URL information from the inner-loop. This will ensure that it is not seen as a way to deploy real workloads in production.
+
+`odo deploy delete` could provide developers with a way to clean up any existing deployment of the application. 
+
+### High-level design:
+
+#### Technical considerations:
+- The implementation would be under the experimental flag.
+- Only supported for Devfile v2.0.0 components.
+- Only supported for Kubernetes/ OpenShift targets.
+- `odo deploy` will not make any changes to the user's project.
+
+#### odo deploy 
+This command will build a container image for the application and deploy it on the target Kubernetes environment. 
+
+Flags:
+ - `--tag`: The tag to be used for the built application container image - `<registry>/<org>/<name>:<tag>` (mandatory).
+ - `service-account`: The service account for running privileged containers and push access to the image registry.
```

Comment:
Is this going to be mandatory? Asking because there's no `--` so it seems more like a positional argument to me. When we talk about `service-account`, aren't we diving into the OpenShift/k8s terminology that we want the odo users to be shielded from?

Reply:
Sorry my mistake. I have fixed this in a later push.

Regarding terminology, I agree we should think of better names. I'm happy if someone what's to suggest a better name for the flags etc.

Reply:
Thinking out loud here.

A simpler alternative could be `user` which `odo` should be smart enough to map to a service-account/user account. Documentation could explain things in depth and it's a good (preferred?) place to use some k8s/openshift lingo.

Reply:
@dharmit Thanks! I like your suggestion. Further discussions with @kadel and @sbose78 meant that we have removed this argument and will use BuildConfig's default account or Kaniko (which des not need a service account. One less argument.. :)